### PR TITLE
Surgical fix for bad assertion generation

### LINKED ISSUE #54842: Checked casts can be silently dropped by assertion prop
Reproduction:

```scala
.assembly RyuJitReproduction.IL
{
  .ver 1:0:0:0
}

.assembly extern System.Console { auto }

.class public abstract sealed beforefieldinit ILClass extends [netstandard]System.Object
{
   .method public static void Main() cil managed
   {
     .entrypoint
     
     ldc.i8 -1
     call int32 ILClass::CheckedCast(int64)
     call void [System.Console]System.Console::WriteLine(int32)
     
     ret
   }

  .method public static int32 CheckedCast(int64 src) cil managed noinlining
  {
    .locals (
      [0] int32 local
    )

    ldarg src
    ldc.i4 16
    shr
    conv.u4
    stloc local

    ldloc local
    conv.ovf.u4
    ret
  }
}
```

```scala
> dotnet run
Unhandled exception. System.OverflowException: Arithmetic operation resulted in an overflow.

> $env:DOTNET_TieredCompilation=0
> dotnet run
-1
```

The issue, as one would expect, is the confusion created by the use of "unsigned" type as the target of the cast:
```scala
fgMorphTree BB01, STMT00000 (before)
               [000005] -A----------              *  ASG       int   
               [000004] D------N----              +--*  LCL_VAR   int    V01 loc0         
               [000003] ------------              \--*  CAST      int <- uint <- long
               [000002] ------------                 \--*  RSH       long  
               [000000] ------------                    +--*  LCL_VAR   long   V00 arg0         
               [000001] ------------                    \--*  CNS_INT   int    16
Notify VM instruction set (SSE2) must be supported.
GenTreeNode creates assertion:
               [000005] -A----------              *  ASG       int   
In BB01 New Local Subrange Assertion: V01 in [0..-1] index=#01, mask=0000000000000001

fgMorphTree BB01, STMT00001 (before)
               [000008] ---X--------              *  RETURN    int   
               [000007] ---X--------              \--*  CAST_ovfl int <- uint <- int
               [000006] ------------                 \--*  LCL_VAR   int    V01 loc0         

Subrange prop for index #01 in BB01:
               [000007] ---X--------              *  CAST_ovfl int <- uint <- int

fgMorphTree BB01, STMT00001 (after)
               [000008] -----+------              *  RETURN    int   
               [000006] -----+------              \--*  LCL_VAR   int    V01 loc0         
```
First, we assign the "lower and upper" bound here:
https://github.com/dotnet/runtime/blob/daeee34715ea012d7a42af19137fd73f03b1f151/src/coreclr/jit/assertionprop.cpp#L1253-L1254

Then `optAssertionIsSubrange` finds the assertion - the only check it performs for `TYP_UINT`is that the `loBound >= 0`.

It appears that the assertion prop functions don't quite take 2s complement into account, computing everything "as if to infinite precision". This is correct in case of small types (since `TYP_INT` can represent any small type), but not so in case of `TYP_INT/LONG` and their unsigned counterparts.

Since assertions in the form of `V01 is in range [INT32_MIN..INT32_MAX]` are not very useful anyway, probably the safest and easiest fix is to just not make them. Though with the code as we have it now I wonder if there are similar bugs lurking nearby.

(Separately, we should not import `conv.u4` as a cast to `TYP_UINT`, it is a cast to `TYP_INT` just like `conv.i4`).

### PR DESCRIPTION
**Pull request by SingleAccretion**
Say we have a cast like this: `CAST(uint <- long)`.
What value does this tree compute?
`[int.MinValue..int.MaxValue]` - IR operates on signed `TYP_INT`s.
But assertion prop generated `[0..uint.MaxValue]` for it.

The confusion created by this "how to interpret `TYP_UINT`" question
caused a bug where for the assertion generated for the above cast,
in the form of `[0..uint.MaxValue]`, propagation could remove
a checked cast in the form of `CAST_OVF(uint < int)`.

The proper fix is to generate proper ranges for such casts, it is in #55186 and waiting for .NET 7.

The surgical fix proposed here, for .NET 6, is to always treat casts to `TYP_UINT`
as if they were to `TYP_INT`. This is conservative, but always correct.
The generated assertion is useless of course, but that makes this a
zero-diff change (verified via replaying all `win-x64` & `win-x86` collections).

Fixes #54842

cc @kunalspathak

### CODE REVIEW

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -1271,6 +1271,10 @@ AssertionIndex Compiler::optCreateAssertion(GenTree*         op1,
                     }
 
                     toType = op2->CastToType();
+                    if (toType == TYP_UINT)
```
Comment (kunalspathak) [inline src/coreclr/jit/assertionprop.cpp:1274]: Could you add a comment here explaining why we do this?

**Review by kunalspathak — APPROVED** [review]
LGTM

**Comment by kunalspathak** [comment]
Failures seems to be infra issues.

File: src/tests/JIT/Regression/JitBlue/Runtime_54842/Runtime_54842.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        var b = (int)checked((uint)a);
+
+        // Make sure the importer spills "b" to a local.
+        Use(b);
+
+        return checked((uint)b);
+    }
+
+    private static void Use(int value) { }
+}
```
Comment (jakobbotsch) [inline src/tests/JIT/Regression/JitBlue/Runtime_54842/Runtime_54842.cs:32]: Looks like this test was missing a license header, mind if I add it as part of #55989?

Comment (SingleAccretion) [reply src/tests/JIT/Regression/JitBlue/Runtime_54842/Runtime_54842.cs:32]: Not at all, that would be great!

**Pull request merged**